Chapter 1 - Introduction (#4)
fsharp/fslang-spec: spec/introduction.md

> # Introduction

Author: It is special static text, see [spec/front-matter.md](https://github.com/fsharp/fslang-spec/blob/main/spec/front-matter.md), which is merged already. (But may still need polishing.)

> # Introduction

Reviewer: for some reason, I cannot mark this resolved...

> F# is a scalable, succinct, type-safe, type-inferred, efficiently executing functional/imperative/object-oriented programming language. It aims to be the premier typed functional programming language for the .NET framework and other implementations of the Ecma 335 Common Language Infrastructure (CLI) specification. F# was partly inspired by the OCaml language and shares some common core constructs with it.

Reviewer 2: I think we want to first merge as is, and then make a branch/tag with new more fresh version, so there are no surprises if we want to review changes, etc.

```fsharp
    let computeDerivative f x =
        let p1 = f (x - 0.05)
      let p2 = f (x + 0.05)
           (p2 - p1) / 0.1
```

The following shows the correct alignment:

```fsharp
    let computeDerivative f x =
        let p1 = f (x - 0.05)
        let p2 = f (x + 0.05)
        (p2 - p1) / 0.1
```

The use of lightweight syntax is the default for all F# code in files with the extension .fs, .fsx, .fsi, or .fsscript.

### Making Data Simple

The first line in our sample simply declares a list of numbers from one through ten.

```fsharp
    let numbers = [1 .. 10]
```

An F# list is an immutable linked list, which is a type of data used extensively in functional programming. Some operators that are related to lists include `::` to add an item to the front of a list and `@` to concatenate two lists. If we try these operators in F# Interactive, we see the following
results:

    > let vowels = ['e'; 'i'; 'o'; 'u'];;
    val vowels: char list = ['e'; 'i'; 'o'; 'u']

    > ['a'] @ vowels;;
    val it: char list = ['a'; 'e'; 'i'; 'o'; 'u']
    
    > vowels @ ['y'];;

>     val it: char list = ['e'; 'i'; 'o'; 'u'; 'y']

Author: I agree it looks nicer as code block, but we have reserved the `fsharp` info string for F# code that we could later extract and verify (like the C# spec does or at least did it).
See also the [conversion guideline](urlhttps://github.com/amplifying-fsharp/fsharplang) that we used.
I will later go separately through the code and turn the fsi examples into code blocks with the `fsother` info string. (As, I believe, I did in later chapters.)

```fsharp
    let rec fib x = if x < 2 then 1 else fib(x-1) + fib(x-2)

    let fibs =
        Async.Parallel [ for i in 0..40 -> async { return fib(i) } ]
        |> Async.RunSynchronously

    printfn "The Fibonacci numbers are %A" fibs

    System.Console.ReadKey(true)
```

The preceding code sample shows multiple, parallel, CPU-bound computations.

F# is also a reactive language. The following example requests multiple web pages in parallel, reacts to the responses for each request, and finally returns the collected results.

```fsharp
    open System
    open System.IO
    open System.Net

    let http url =
        async {
            let req = WebRequest.Create(Uri url)
            use! resp = req.AsyncGetResponse()
            use stream = resp.GetResponseStream()
            use reader = new StreamReader(stream)
            let contents = reader.ReadToEnd()
            return contents
            }

    let sites = [
        "http://www.bing.com"
        "http://www.google.com"
        "http://www.yahoo.com"
        "http://www.search.com"
        ]

    let htmlOfSites =
        Async.Parallel [for site in sites -> http site ]
        |> Async.RunSynchronously

> ```

Author: Yes, Fantomas makes sense. But I would like to do that then in an automated way. See also my above remark of using code blocks with the `fsharp` info string only for F# code.

By using asynchronous workflows together with other CLI libraries, F# programs can implement parallel tasks, parallel I/O operations, and message-receiving agents.

### Strong Typing for Floating-Point Code

> F# applies type checking and type inference to floating-point-intensive domains through units of measure inference and checking. This feature allows you to type-check programs that manipulate floating-point numbers that represent physical and abstract quantities in a stronger way than other typed languages, without losing any performance in your compiled code. You can think of this feature as providing a type system for floating-point code.

Reviewer: ```suggestion
F# applies type checking and type inference to floating-point-intensive domains through *units of measure inference and checking*. This feature allows you to type-check programs that manipulate floating-point numbers that represent physical and abstract quantities in a stronger way than other typed languages, without losing any performance in your compiled code. You can think of this feature as providing a type system for floating-point code.
```

Consider the following example:

```fsharp
    [<Measure>] type kg
    [<Measure>] type m
    [<Measure>] type s

    let gravityOnEarth = 9.81<m/s^2>
    let heightOfTowerOfPisa = 55.86<m>
    let speedOfImpact = sqrt(2.0 * gravityOnEarth * heightOfTowerOfPisa)
```

The `Measure` attribute tells F# that `kg`, `s`, and `m` are not really types in the usual sense of the word, but are used to build units of measure. Here `speedOfImpact` is inferred to have type `float<m/s>`.

### Object-Oriented Programming and Code Organization

> The sample program shown at the start of this chapter is a _script_. Although scripts are excellent for rapid prototyping, they are not suitable for larger software components. F# supports the transition from scripting to structured code through several techniques.

Reviewer: We can use single style for markdown in such a large body of work (VSCode can do this), and we can make this a lint-check in PRs as well perhaps? Not sure how precise we should be here.
```suggestion
The sample program shown at the start of this chapter is a *script*. Although scripts are excellent for rapid prototyping, they are not suitable for larger software components. F# supports the transition from scripting to structured code through several techniques.
```

Author: I agree, we should use a single style

```fsharp
    module ApplicationLogic =
        let numbers n = [1 .. n]
        let square x = x * x
        let squares n = numbers n |> List.map square

    printfn "Squares up to 5 = %A" (ApplicationLogic.squares 5)
    printfn "Squares up to 10 = %A" (ApplicationLogic.squares 10)
    System.Console.ReadKey(true)
```

Modules are also used in the F# library design to associate extra functionality with types. For example, `List.map` is a function in a module.

Other mechanisms aimed at supporting software engineering include *signatures*, which can be used to give explicit types to components, and namespaces, which serve as a way of organizing the name hierarchies for larger APIs.

### Information-rich Programming

F# Information-rich programming addresses the trend toward greater availability of data, services, and information. The key to information-rich programming is to eliminate barriers to working with diverse information sources that are available on the Internet and in modern enterprise environments. Type providers and query expressions are a significant part of F# support for information-rich programming.

> The F# Type Provider mechanism allows you to seamlessly incorporate, in a strongly typed manner, data and services from external sources. A type provider presents your program with new types and methods that are typically based on the schemas of external information sources. For example, an F# type provider for Structured Query Language (SQL) supplies types and methods that allow programmers to work directly with the tables of any SQL database:

Reviewer: ```suggestion
The F# Type Provider mechanism allows you to seamlessly incorporate, in a strongly typed manner, data and services from external sources. A *type provider* presents your program with new types and methods that are typically based on the schemas of external information sources. For example, an F# type provider for Structured Query Language (SQL) supplies types and methods that allow programmers to work directly with the tables of any SQL database:
```

```fsharp
    let countOfCustomers =
        query {
            for customer in db.Customers do
                where (customer.LastName.StartsWith("N"))
                select (customer.FirstName, customer.LastName)
            }
```

Now it is easier than ever to access many important data sources—including enterprise, web, and cloud—by using a set of built-in type providers for SQL databases and web data protocols. Where necessary, you can create your own custom type providers or reference type providers that others have created. For example, assume your organization has a data service that provides a large and growing number of named data sets, each with its own stable data schema. You may choose to create a type provider that reads the schemas and presents the latest available data sets to the programmer in a strongly typed way.

## Notational Conventions in This Specification

This specification describes the F# language by using a mixture of informal and semiformal techniques. All examples in this specification use lightweight syntax, unless otherwise specified.

Regular expressions are given in the usual notation, as shown in the table:

> | Notation | Meaning |
> | --- | --- |
> | regexp+ | One or more occurrences |
> | regexp* | Zero or more occurrences |
> | regexp? | Zero or one occurrences |
> | [ char - char ] | Range of ASCII characters |
> | [ ^ char - char ] | Any characters except those in the range |

Reviewer: Code readability (in VSCode, I use 'Markdown all in one' to format a block)
```suggestion
| Notation          | Meaning                                  |
| ----------------- | ---------------------------------------- |
| regexp+           | One or more occurrences                  |
| regexp*           | Zero or more occurrences                 |
| regexp?           | Zero or one occurrences                  |
| [ char - char ]   | Range of ASCII characters                |
| [ ^ char - char ] | Any characters except those in the range |
```

Unicode character classes are referred to by their abbreviation as used in CLI libraries for regular expressions—for example, `\Lu` refers to any uppercase letter. The following characters are referred to using the indicated notation:

> | Character | Name | Notation |
> | --- | --- | --- |
> | \b | backspace | ASCII/UTF-8/UTF-16/UTF-32 code 08 |
> | \n | newline | ASCII/UTF-8/UTF-16/UTF-32 code 10 |
> | \r | return | ASCII/UTF-8/UTF-16/UTF-32 code 13 |
> | \t | tab | ASCII/UTF-8/UTF-16/UTF-32 code 09 |

Reviewer: Same:
```suggestion
| Character | Name      | Notation                          |
| --------- | --------- | --------------------------------- |
| \b        | backspace | ASCII/UTF-8/UTF-16/UTF-32 code 08 |
| \n        | newline   | ASCII/UTF-8/UTF-16/UTF-32 code 10 |
| \r        | return    | ASCII/UTF-8/UTF-16/UTF-32 code 13 |
| \t        | tab       | ASCII/UTF-8/UTF-16/UTF-32 code 09 |
```

Strings of characters that are clearly not a regular expression are written verbatim. Therefore, the following string

    abstract

matches precisely the characters `abstract`.

Where appropriate, apostrophes and quotation marks enclose symbols that are used in the specification of the grammar itself, such as `'<'` and `'|'`. For example, the following regular expression matches `(+)` or `(-)`:

    '(' (+|-) ')'

This regular expression matches precisely the characters `#if`:

    "#if"

Regular expressions are typically used to specify tokens.

    token token-name = regexp

> In the grammar rules, the notation `element-name~opt` indicates an optional element. The notation `...` indicates repetition of the preceding non-terminal construct and the separator token. For example, `expr ',' ... ',' expr` means a sequence of one or more `expr` elements separated by commas.

Reviewer: maybe:
```suggestion
In the grammar rules, the notation `element-nameₒₚₜ` indicates an optional element. The notation `...` indicates repetition of the preceding non-terminal construct and the separator token. For example, `expr ',' ... ',' expr` means a sequence of one or more `expr` elements separated by commas.
```

Reviewer: Per conclusion in that discussion: it is now a question mark (`?`).

Other mechanisms aimed at supporting software engineering include *signatures*, which can be used to give explicit types to components, and namespaces, which serve as a way of organizing the name hierarchies for larger APIs.

### Information-rich Programming

F# Information-rich programming addresses the trend toward greater availability of data, services, and information. The key to information-rich programming is to eliminate barriers to working with diverse information sources that are available on the Internet and in modern enterprise environments. Type providers and query expressions are a significant part of F# support for information-rich programming.

The F# Type Provider mechanism allows you to seamlessly incorporate, in a strongly typed manner, data and services from external sources. A *type provider* presents your program with new types and methods that are typically based on the schemas of external information sources. For example, an F# type provider for Structured Query Language (SQL) supplies types and methods that allow programmers to work directly with the tables of any SQL database:

```fsharp

> // Add References to FSharp.Data.TypeProviders, System.Data, and System.Data.   Linq
> type schema = SqlDataConnection<"Data Source=localhost;Integrated   Security=SSPI;">

Reviewer: nit:
```suggestion
// Add References to FSharp.Data.TypeProviders, System.Data, and System.Data.Linq
type schema = SqlDataConnection<"Data Source=localhost;Integrated Security=SSPI;">
```

| Notation          | Meaning                                  |
| ----------------- | ---------------------------------------- |
| regexp+           | One or more occurrences                  |
| regexp*           | Zero or more occurrences                 |
| regexp?           | Zero or one occurrences                  |
| [ char - char ]   | Range of ASCII characters                |
| [ ^ char - char ] | Any characters except those in the range |

Unicode character classes are referred to by their abbreviation as used in CLI libraries for regular expressions—for example, `\Lu` refers to any uppercase letter. The following characters are referred to using the indicated notation:

| Character | Name      | Notation                          |
| --------- | --------- | --------------------------------- |
| \b        | backspace | ASCII/UTF-8/UTF-16/UTF-32 code 08 |
| \n        | newline   | ASCII/UTF-8/UTF-16/UTF-32 code 10 |
| \r        | return    | ASCII/UTF-8/UTF-16/UTF-32 code 13 |
| \t        | tab       | ASCII/UTF-8/UTF-16/UTF-32 code 09 |

Strings of characters that are clearly not a regular expression are written verbatim. Therefore, the following string

>     abstract

Reviewer: Not necessary now, but to recognize these blocks automatically, perhaps we can start using `abnf` sections (even though it isn't exactly abnf)?
`````suggestion
```abnf
abstract
```
`````

matches precisely the characters `abstract`.

Where appropriate, apostrophes and quotation marks enclose symbols that are used in the specification of the grammar itself, such as `'<'` and `'|'`. For example, the following regular expression matches `(+)` or `(-)`:

>     '(' (+|-) ')'

Reviewer: Same here:
`````suggestion
```abnf
'(' (+|-) ')'
```
`````

This regular expression matches precisely the characters `#if`:

>     "#if"

Reviewer: `````suggestion
```abnf
"#if"
```
`````

Regular expressions are typically used to specify tokens.

>     token token-name = regexp

Reviewer: `````suggestion
```abnf
token token-name = regexp
```
`````
